### ISSUE TITLE
Method implementation is incorrectly reported as having inconsistent signature with one of the overloads

### ISSUE THREAD

**Issue by danpascu** [microsoft/pyright#8110]
When I use an enum value as the default value for a method argument, pyright incorrectly flags the method implementation as being inconsistent with its overloads.

With this piece of code (the parts not relevant to this were removed for brevity):

```py
from collections.abc import MutableMapping
from enum import Enum
from typing import overload, reveal_type


class Marker(Enum):
    MissingArgument = 'Argument is not provided'


class weakobjectmap[K, V](MutableMapping[K, V]):
    __data__: dict[int, V]  # this is initialized to {} in __init__

    ...

    @overload
    def pop(self, key: K, /) -> V: ...

    @overload
    def pop[T](self, key: K, /, default: T) -> V | T: ...

    def pop(self, key, /, default=Marker.MissingArgument):
        try:
            if default is Marker.MissingArgument:
                return self.__data__.pop(id(key))
            return self.__data__.pop(id(key), default)
        except KeyError:
            raise KeyError(key) from None
```

pyright reports the following error:

```
weakref.py
  weakref.py:230:9 - error: Overloaded implementation is not consistent with signature of overload 2
    Function return type "V@weakobjectmap | T@pop" is incompatible with type "V@weakobjectmap"
      Type "V@weakobjectmap | T@pop" is incompatible with type "V@weakobjectmap" (reportInconsistentOverload)
1 error, 0 warnings, 0 informations
```

If instead of the enum Marker I use an object marker like `MissingArgument = object()` then pyright accepts it and reports no error. At the same time I also checked with mypy and mypy accepts both without error.

I believe the issue originates because the output of `self.__data__.pop(...)` is not identified correctly. If I slightly modify the code like this (by adding `reveal_type(...)` around the returned values):

```py
    def pop(self, key, /, default=Marker.MissingArgument):
        try:
            if default is Marker.MissingArgument:
                return reveal_type(self.__data__.pop(id(key)))
            return reveal_type(self.__data__.pop(id(key), default))
        except KeyError:
            raise KeyError(key) from None
```

then pyright reports this in addition to the error:

```
  weakref.py:233:36 - information: Type of "self.__data__.pop(id(key))" is "V@weakobjectmap"
  weakref.py:234:32 - information: Type of "self.__data__.pop(id(key), default)" is "V@weakobjectmap"
```

While the 1st is correct, the second looks incorrect as it should have been `"V@weakobjectmap | T@pop"` or maybe `"V@weakobjectmap | object"` or just `Unknown`.

At the same time mypy seems to correctly identify the output as:

```
weakref.py:233: note: Revealed type is "V`2"
weakref.py:234: note: Revealed type is "Any"
```

If I use the object marker as mentioned above, then the second output is correctly identified as `"V@weakobjectmap | object"` and the error goes away.

This might be happening because pyright sees `Marker.MissingArgument` as `Literal[Marker.MissingArgument]`, not as an instance of `Marker`. But regardless of how it's seen, it should not matter because the marker is never involved in the output. Only when default is not `Marker.MissingArgument` does the output involve the type of the `default` argument.

This happens with the command line pyright version 1.1.366

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

Unlike mypy, pyright infers return types if a return type annotation is omitted. In this case, it is inferring the type `V`, which is not consistent with the second overload signature. 

You have two options here:
1. To replicate mypy's behavior, you can provide an explicit `-> Any` return type annotation.
2. If you want pyright to perform complete type checking of the overload implementation, supply a fully-annotated signature:
```python
    def pop[T](self, key: K, /, default: T | Literal[Marker.MissingArgument]=Marker.MissingArgument) -> V | T: ...
```

**Comment by danpascu** [comment]
I have to disagree. I do not think that pyright is working as intended when it infers the return type as V only on that branch.

According to the signature for the dict.pop method (the appropriate overload) we have:

`dict[K, V].pop[T](key: K, default: T) -> V | T: ...`

On that branch we call `dict[K, V].pop[T](key: K, default: T)` which according to the overload signature cannot have its return type being inferred as V only, since if the key is missing `default` will be returned and default is of type T.

What I have an issue with is that both of these:

`dict[K, V].pop(key: K)`

and

`dict[K, V].pop[T](key: K, default: T)`

are inferred to only return V as demonstrated by the 2 reveal_type calls inserted on the returned values.

This is not about replication mypy's behavior.

**Comment by erictraut** [comment]
For information about how annotations affect type inference, refer to the documentation about [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types).

Signatures marked as `@overload` are used only in the evaluation of calls to the overloaded function. They are never used for type checking the overload implementation. The implementation should have its own annotations. Type checkers like pyright and mypy will then validate that the annotated implementation is compatible with all of the overloaded signatures. That's how overloads are defined to work in Python's type system and in other languages like TypeScript that offer overloads.

For information about the `*`, refer to the documentation about [conditional types](int | float | bool).

**Comment by danpascu** [comment]
Thanks for the pointers.

Interestingly enough, I used to fully annotate both the overloads and the actual implementation, until I noticed that in a lot of projects, other people only annotated the overloads and provided 0 annotations for the actual implementation. So I wrongly concluded that they were not necessary. I also ran into a case where I had a lot of trouble annotation the actual implementation, but with annotations removed it worked without a problem. That led me to believe that maybe implementation annotations only complicate the matter and are unnecessary once you defined all the overloads, and that's why other people seem to leave them out, so I started doing the same, thinking that maybe that's the recommended way of doing it.

Again thanks for the pointers.

**Comment by erictraut** [comment]
It's fine to omit the annotations for the implementation if you are a library author who doesn't care about type checking your own code but want to improve the experience for consumers of your library. That's perhaps where you've seen this pattern.
